src/Microsoft.Framework.OptionsModel/ConfigurationBinder.cs

```diff
@@ -0,0 +1,290 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Framework.ConfigurationModel;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.Framework.OptionsModel
+{
+    public class ConfigurationBinder
+    {
+        private readonly Stack<IConfiguration> _context;
+        private readonly IConfiguration _startConfiguration;
+
+        private IConfiguration _currentConfigurationContext;
+
+        public ConfigurationBinder([NotNull] IConfiguration configuration)
+        {
+            _context = new Stack<IConfiguration>();
+            _startConfiguration = configuration;
+        }
+
+        public static void Bind(IConfiguration configuration, object merge)
+        {
+            new ConfigurationBinder(configuration).Read(merge);
+        }
+
+        public void Read(object merge)
+        {
+            if (merge == null)
+            {
+                return;
+            }
+
+            _context.Clear();
+            EnterContext(_startConfiguration);
+            VisitObject(merge);
+            ExitContext();
+        }
+
+        public TOptions Read<TOptions>() where TOptions : new()
+        {
+            var options = new TOptions();
+            Read(options);
+            return options;
+        }
+
+        private void VisitObject(object obj)
+        {
+            foreach (var property in GetAllProperties(obj.GetType().GetTypeInfo()))
+            {
+                VisitProperty(property, obj);
+            }
+        }
+
+        private void VisitProperty(PropertyInfo property, object propertyOwner)
+        {
+            EnterContext(_currentConfigurationContext.GetSubKey(property.Name));
+            try
+            {
+                if (property.GetMethod == null || !property.GetMethod.IsPublic)
+                {
+                    // We don't support set only properties
+                    return;
+                }
+
+                var propertyValue = property.GetValue(propertyOwner);
+                var hasPublicSetter = property.SetMethod != null && property.SetMethod.IsPublic;
+
+                if (propertyValue == null && !hasPublicSetter)
+                {
+                    // Property doesn't have a value and we cannot set it so there is no
+                    // point in going further down the graph
+                    return;
+                }
+
+                VisitType(property.PropertyType, ref propertyValue);
```

Comment:
Yeah I think that is normally understood that objects are mutable.  I don't think we typically use ref parameters in general :)  It definitely sticks out a bit like a sore thumb... I'd just call it out in a comment for the method if you want to emphasize this point.  Making the signature void/ref is overkill (especially since this is a private method)...